File: trees/cppRealizations/binaryTreeLevelOrderTraversal.cpp
#include "btf.h"
#include <vector>
#include <queue>

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root == nullptr) return {};
        vector<vector<int>> result = {{root->val}};
        deque<TreeNode*> q = {root};
        while(!q.empty()){
            int size = q.size();
            vector<int> level;
            for(int i = 0; i < size; i++){
                TreeNode* node = q.front();
                q.pop_front();
                if(node->left != nullptr){
                    q.push_back(node->left);
                    level.push_back(node->left->val);
                }
                if(node->right != nullptr){
                    q.push_back(node->right);
                    level.push_back(node->right->val);
                }
            }
            if(level.size() > 0){
                result.push_back(level);
            }
        }
        return result;
        
    }
};


int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    cout << "Input:" << endl;
    btf(root);
    cout << "Solution:" << endl;
    vector<vector<int>> result = Solution().levelOrder(root);
    
    cout << "[ ";
    for(auto vec : result){
        cout << "[ ";
        for(auto item : vec){
            cout << item << ", ";
        }
        cout << "], ";
    }
    cout << endl;
    return 0;
}

File: trees/binaryTreeLevelOrderTraversal.py
from trees.btf import *


class Solution:
    def levelOrder(self, root: TreeNode) -> list[list[int]]:
        if not root:
            return []
        result = [[root.val]]
        q = deque([root])
        while q:
            level = []
            for i in range(len(q)):
                node = q.popleft()
                if node.left:
                    level.append(node.left.val)
                    q.append(node.left)
                if node.right:
                    level.append(node.right.val)
                    q.append(node.right)
            if level:
                result.append(level)
        return result




if __name__ == '__main__':
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print("Input:")
    btf(root)
    print("Sulution:")
    print(Solution().levelOrder(root))